File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 * int val;
 * TreeNode *left;
 * TreeNode *right;
 * TreeNode() : val(0), left(nullptr), right(nullptr) {}
 * TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <vector>
#include <queue>

class Solution {
public:
    std::vector<std::vector<int>> levelOrder(TreeNode* root) {
        std::vector<std::vector<int>> result;
        
        // Edge case: If the tree is completely empty, return the empty result
        if (root == nullptr) {
            return result;
        }
        
        std::queue<TreeNode*> q;
        q.push(root);
        
        // Continue traversing as long as there are nodes in the queue
        while (!q.empty()) {
            int level_size = q.size(); // Snapshot the number of nodes at this specific level
            std::vector<int> current_level;
            
            // Process ONLY the nodes that belong to the current level
            for (int i = 0; i < level_size; ++i) {
                TreeNode* node = q.front();
                q.pop();
                
                current_level.push_back(node->val);
                
                // Queue up the next level's nodes
                if (node->left != nullptr) {
                    q.push(node->left);
                }
                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }
            
            // Append the fully processed level to our final result
            result.push_back(current_level);
        }
        
        return result;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        curr = root
        
        # We loop as long as there are nodes to process in the stack, 
        # or we have a current node we are actively looking at.
        while stack or curr:
            
            # Keep diving left until we hit a dead end (null)
            # This ensures we always reach the absolute smallest remaining number
            while curr:
                stack.append(curr)
                curr = curr.left
                
            # Pop the top node from the stack (the next smallest number)
            curr = stack.pop()
            
            # We "visited" a node in sorted order, so we decrement our counter
            k -= 1
            
            # If k is 0, we've found our kth smallest number
            if k == 0:
                return curr.val
                
            # After processing the node, we must check if it has any numbers
            # on its right side (which are larger than it, but smaller than its parent)
            curr = curr.right
            
        return -1 # Fallback, though the problem guarantees k is valid